libraries/AP_ExternalAHRS/AP_ExternalAHRS.cpp

```diff
@@ -0,0 +1,560 @@
+    gps.vdop = pkt2.GPS1DOP[3];
+
+    gps.latitude = pkt2.GPS1posLLA[0] * 1.0e7;
+    gps.longitude = pkt2.GPS1posLLA[1] * 1.0e7;
+    gps.msl_altitude = pkt2.GPS1posLLA[2] * 1.0e2;
+
+    gps.ned_vel_north = pkt2.GPS1velNED[0];
+    gps.ned_vel_east = pkt2.GPS1velNED[1];
+    gps.ned_vel_down = pkt2.GPS1velNED[2];
+
+    if (gps.fix_type >= 3 && !origin_set) {
+        origin.lat = gps.latitude;
+        origin.lng = gps.longitude;
+        origin.alt = gps.msl_altitude;
+        origin_set = true;
+    }
+
+    AP::gps().handle_external(gps);
+}
+
+// get serial port number for the uart
+int8_t AP_ExternalAHRS::get_port(void) const
+{
+    if (!uart) {
+        return -1;
+    }
+    return port_num;
+};
+
+// accessors for AP_AHRS
+bool AP_ExternalAHRS::healthy(void) const
+{
+    uint32_t now = AP_HAL::millis();
+    return (now - last_pkt1_ms < 40 && now - last_pkt2_ms < 500);
+}
+
+bool AP_ExternalAHRS::initialised(void)
+{
+    return last_pkt1_ms != 0 && last_pkt2_ms != 0;
+}
+
+bool AP_ExternalAHRS::get_quaternion(Quaternion &quat)
+{
+    if (!last_pkt1) {
+        return false;
+    }
+    WITH_SEMAPHORE(sem);
+    const struct VN_packet1 &pkt1 = *last_pkt1;
+    quat(pkt1.quaternion[3], pkt1.quaternion[0], pkt1.quaternion[1], pkt1.quaternion[2]);
+    return true;
+}
+
+bool AP_ExternalAHRS::get_origin(Location &loc)
+{
+    WITH_SEMAPHORE(sem);
+    loc = origin;
+    return origin_set;
+}
+
+bool AP_ExternalAHRS::get_location(Location &loc)
+{
+    if (!last_pkt2) {
+        return false;
+    }
+    const struct VN_packet2 &pkt2 = *last_pkt2;
+    WITH_SEMAPHORE(sem);
+    loc = Location();
+    loc.lat = pkt2.GPS1posLLA[0] * 1.0e7;
+    loc.lng = pkt2.GPS1posLLA[1] * 1.0e7;
+    loc.alt = pkt2.GPS1posLLA[2] * 1.0e2;
+    return true;
+}
+
+Vector2f AP_ExternalAHRS::get_groundspeed_vector()
+{
+    if (!last_pkt1) {
+        return Vector2f();
+    }
+    const struct VN_packet1 &pkt1 = *last_pkt1;
+    WITH_SEMAPHORE(sem);
+    return Vector2f(pkt1.velNED[0], pkt1.velNED[1]);
+}
+
+bool AP_ExternalAHRS::get_velocity_NED(Vector3f &vel)
+{
+    if (!last_pkt1) {
+        return false;
+    }
+    const struct VN_packet1 &pkt1 = *last_pkt1;
+    WITH_SEMAPHORE(sem);
+    vel.x = pkt1.velNED[0];
+    vel.y = pkt1.velNED[1];
+    vel.z = pkt1.velNED[2];
+    return true;
+}
+
+bool AP_ExternalAHRS::get_speed_down(float &speedD)
+{
+    if (!last_pkt1) {
+        return false;
+    }
+    const struct VN_packet1 &pkt1 = *last_pkt1;
+    WITH_SEMAPHORE(sem);
+    speedD = pkt1.velNED[2];
+    return true;
+}
+
+bool AP_ExternalAHRS::pre_arm_check(char *failure_msg, uint8_t failure_msg_len) const
+{
+    if (!healthy()) {
+        hal.util->snprintf(failure_msg, failure_msg_len, "VectorNav unhealthy");
+        return false;
+    }
+    if (last_pkt2->GPS1Fix < 3) {
+        hal.util->snprintf(failure_msg, failure_msg_len, "VectorNav no GPS1 lock");
+        return false;
+    }
+    if (last_pkt2->GPS1Fix < 3) {
```

Comment:
```suggestion
    if (last_pkt2->GPS2Fix < 3) {
```

Only guessing here :-)

Reply:
fixed